driver-core/src/main/com/mongodb/connection/netty/NettyStream.java

```diff
@@ -211,18 +269,32 @@ public void operationComplete(final ChannelFuture future) throws Exception {
 
     @Override
     public void readAsync(final int numBytes, final AsyncCompletionHandler<ByteBuf> handler) {
-        readAsync(numBytes, handler, 0);
+        readAsync(numBytes, handler, readTimeoutMillis);
     }
 
-    private void readAsync(final int numBytes, final AsyncCompletionHandler<ByteBuf> handler, final int additionalTimeout) {
-        scheduleReadTimeout(additionalTimeout);
+    /**
+     * @param numBytes Must be equal to {@link #pendingReader}{@code .numBytes} when called by a Netty channel handler.
+     * @param handler Must be equal to {@link #pendingReader}{@code .handler} when called by a Netty channel handler.
+     * @param readTimeoutMillis Must be equal to {@link #NO_SCHEDULE_TIMEOUT} when called by a Netty channel handler.
+     *                          Timeouts may be scheduled only by the public read methods. Taking into account that concurrent pending
+     *                          readers are not allowed, there must not be a situation when threads attempt to schedule a timeout
+     *                          before the previous one is either cancelled or completed.
+     */
+    private void readAsync(final int numBytes, final AsyncCompletionHandler<ByteBuf> handler, final long readTimeoutMillis) {
         ByteBuf buffer = null;
         Throwable exceptionResult = null;
         synchronized (this) {
             exceptionResult = pendingException;
+            PendingReader pendingReader = this.pendingReader;
             if (exceptionResult == null) {
                 if (!hasBytesAvailable(numBytes)) {
-                    pendingReader = new PendingReader(numBytes, handler);
+                    if (pendingReader == null) {//called by a public read method
+                        pendingReader = new PendingReader(numBytes, handler, scheduleReadTimeout(readTimeoutTask, readTimeoutMillis));
+                        this.pendingReader = pendingReader;
+                    } else {//called by a Netty channel handler
```

Comment:
This else statement can be removed

Reply:
I agree that it can be, but does removing it make the code clearer? This code is not so trivial, and having comments that express what's going on, or even state some invariants and possible interpretations in case they are violated, appears to be useful. What are the downsides of having code comments like this?

Reply:
We generally don't inline comment in the code because its considered a code smell.  The code should be readable in intent, however there are scenarios where you may want to add comments. In general methods / classes have comments describing the logic. 

I'll leave this one to you thinks best, It may be useful in x years when we revisit this code.

Reply:
Regarding this else statement I don't think it adds to the code as contains no logic.

A single comment checking is `if(pendingReader == null)` should suffice.

Reply:
If you really want the commented asserts to remain as documentation, how about also commenting out the `else` that contains them?

Reply:
Removed the `else` block and the comments within.